fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,124 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    /**
+     * Use fetch to send a GET request to `/api/todos`.
+     * Then 😉, once the response JSON is received and parsed,
+     * update state with the received todos.
+     */
+    fetch('/api/todos', {
+      method: 'GET'
+    })
+      .then(response => response.json())
+      .then(data => {
+        this.setState({ todos: data });
+      })
+      .catch(error => {
+        console.error('Error:', error);
+      });
+  }
+
+  addTodo(newTodo) {
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * Do not mutate the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(response => response.json())
+      .then(newTodo => {
+        this.setState({ todos: this.state.todos.concat(newTodo) });
+      })
+      .catch(error => {
+        console.error('Error:', error);
+      });
+  }
+
+  toggleCompleted(todoId) {
+    const newObject = {};
+    let index = 0;
+    for (let i = 0; i < this.state.todos.length; i++) {
+      if (todoId === this.state.todos[i].todoId) {
+        newObject.isCompleted = !this.state.todos[i].isCompleted;
+        index = i;
+        break;
+      }
+    }
+    fetch(`/api/todos/${todoId}`, {
+      method: 'PATCH',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newObject)
+    })
+      .then(response => response.json())
+      .then(data => {
+        const newArray = this.state.todos;
```

Comment:
Arrays are reference data types so this won't copy the array it will just create a new pointer to the same array. You can use `slice` or the ES6 spread operator to create a shallow copy of the array.

```JAVASCRIPT
// Using .slice
const newArray = this.state.todos.slice();

// Using ES6 spread operator
const newArray = [ ...this.state.todos ];
```

Reply:
I hope my changes meets your standards oh great one.
![fetch-in-react-practice final turn in](https://user-images.githubusercontent.com/71291742/139774593-bfff03bf-02bc-48ff-8ced-96de32c46467.gif)